eloquentjs/4_3_list.js

```diff
@@ -0,0 +1,119 @@
+/**
+ * Converts an array  to a list data structure
+ * @param {*} inputArray 
+ * @param {*} list 
+ * @returns 
+ */
+function arrayToList([...inputArray], list = { value: null, rest: null }) { 
+    if (!Array.isArray(inputArray)) { //this needs work
+        throw "input must be array";
+    }
+
+    if (inputArray.length === 0) {
+        return list.rest;
+    }
+
+    list.value = inputArray.pop();
+    list.rest = structuredClone(list);
+
+    return arrayToList(inputArray, list);
+}
+
+/**
+ * Converts a list data structure to an array
+ * @param {*} inputList 
+ * @param {*} arrayAccumulator 
+ * @returns 
+ */
+function listToArray(inputList, arrayAccumulator = []) {
+    if (!Object.keys(inputList).includes('value') || !Object.keys(inputList).includes('rest')) {
+        throw "input must be list data structure";
+    }
+    
+    arrayAccumulator.push(inputList.value);
+
+    if (inputList.rest) {
+        listToArray(inputList.rest,arrayAccumulator)
+    }
+
+    return arrayAccumulator;
+}
+
+/**
+ * Returns a copy of a list with a prepended element
+ * @param {*} inputList 
+ * @param {*} element 
+ * @returns 
+ */
+function prepend(inputList, element) {
+    if (!Object.keys(inputList).includes('value') || !Object.keys(inputList).includes('rest')) {
+        throw "input must be list data structure";
+    }
+
+    return {
+        value: element,
+        rest: inputList
+    }
+}
+
+/**
+ * Returns the nth element of a list
+ * @param {*} inputList 
+ * @param {*} targetIndex 
+ * @param {*} depth 
+ * @returns 
+ */
+function nth(inputList, targetIndex, depth = 0) {
```

Comment:
You can do this without the depth argument. Try to figure out how. Also, really think about your termination condition. It can be made a good bit simpler.